feat: Add HRM Model

zbloss opened the pull request:
# What does this PR do?

Adds the new Hierarchical Reasoning Model.

Fixes #41271

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker @Cyrilvallez hoping you can take a look. The model is currently implemented as a text model though further research hints it may be helpful for understanding images as well.

zbloss committed:
feat: Add HRM Model

zbloss committed:
fix linting issues

zbloss committed:
fix lint and doc issues

zbloss committed:
fix attention and rope tests

zbloss committed:
fix linting error

zbloss committed:
fixed

zbloss committed:
check modular conversion now passing

zbloss committed:
add correct md model link

zbloss committed:
fix check config attribute related issues

molbap reviewed (commented):
Done a shallow first pass to leave some pointers - feel free to ask questions back and congrats for this endeavour! Exciting!

zbloss commented on src/transformers/models/hrm/modular_hrm.py:
```diff
@@ -0,0 +1,972 @@
+# coding=utf-8
+# Copyright 2025 The HRM Team and The HuggingFace Inc. team. All rights reserved.
```
Yep good call, I shorthanded this

zbloss committed:
fix weird checkpoint issues

zbloss committed:
remove test fix formatting

zbloss commented:
@molbap I believe I fixed everything, let me know when you get a chance to take another look.

I converted the author's original weights and uploaded them here after verifying that they work: 
* https://huggingface.co/zbloss/HRM-maze-30x30-hard
* https://huggingface.co/zbloss/HRM-sudoku-extreme

molbap reviewed (commented):
Thanks for iterating! left additional comments

zbloss committed:
add convert_original_hrm_weights_to_hf script

reuse components in module_hrm.py

improve tests by using actual checkpoint

remove unused tests

zbloss commented on tests/models/hrm/test_modeling_hrm.py:
```diff
@@ -0,0 +1,474 @@
+        # This test requires the converted checkpoint to be available
+        # Skip if the model is not yet available on the Hub
+        try:
+            model = HrmForCausalLM.from_pretrained("zbloss/HRM-sudoku-extreme")
```
Yes, adding the conversion script

zbloss committed:
addressing feedback

zbloss committed:
regenerated code

zbloss committed:
update doc copyright

molbap reviewed (commented):
Left several comments - one very important is that currently there is no way to _actually_ test the model, and that's a bigger blocker than the rest. When you have an end-to-end test (using the weights you converted) we'll be able to move a bit faster.

The rest is more about being inline with the rest of the library. 
- Variable names should be explicit.
- we should avoid if/else codepaths in the modeling. For instance for rope, it should be clearly defined in one place.
- for init_weights, I understand the truncated normal init schema is needed, but it should be interfaced in a familiar way (especially for an "exotic" model with many new components, "usual" components have an even bigger motivation to stay, well, usual.

Finally, please let's only resolve the conversations when a reviewer had the time to check it or it's really obvious, else it makes it hard for us to track iterations :pray: 

Thanks and let me know if you face issues!

zbloss committed:
addressing feedback

zbloss committed:
fixing failing tests

zbloss committed:
ixing torch_tests

zbloss committed:
fix issue regarding -5.0 bias

molbap commented on tests/models/hrm/test_modeling_hrm.py:
```diff
@@ -0,0 +1,704 @@
+                outputs = model(input_ids=partial_puzzle, state=state)
+                state = outputs.carry
+
+                # Check if halted
```
these comments are not needed, as the code they comment is self-explanatory
```suggestion
```

molbap commented on tests/models/hrm/test_modeling_hrm.py:
```diff
@@ -0,0 +1,704 @@
+
+        self.assertIsNotNone(generated)
+        self.assertEqual(generated.shape[0], 1)
+        self.assertGreater(generated.shape[1], input_ids.shape[1])
```
This does not test generation capabilities, it just tests that "something" is generated - it's not an integration test

molbap commented on src/transformers/models/hrm/convert_original_hrm_weights_to_hf.py:
```diff
@@ -0,0 +1,188 @@
+            config_dict = json.load(f)
+        config = HrmConfig(**config_dict)
+    else:
+        # Try to extract config from checkpoint metadata
```
we can do without all these comments
```suggestion
```

molbap commented on src/transformers/models/hrm/modular_hrm.py:
```diff
@@ -0,0 +1,1017 @@
+
+    def forward(self, position_ids=None) -> tuple[torch.Tensor, torch.Tensor]:
+        """Forward pass - position_ids parameter is accepted for compatibility but not used."""
+        return self.cos_cached, self.sin_cached
```
agree it's simpler but we're going to overhaul RoPE soon, for the time being let's stick to the existing design please!

molbap commented on src/transformers/models/hrm/modular_hrm.py:
```diff
@@ -0,0 +1,1017 @@
+        output = torch.matmul(attn, value)
+
+        # Transpose back to (batch, seq_len, num_heads, head_dim)
+        return output.transpose(1, 2).contiguous()
```
This should be a helper function outside of the class, and aligned with existing ` eager_attention_forward`, why not use directly the one from llama for instance?

molbap reviewed (commented):
Left another review with some points to improve, thanks!

zbloss committed:
addressing feedback

zbloss committed:
Merge branch 'main' into main

zbloss committed:
fix_and_overwrite

molbap commented on tests/models/hrm/test_modeling_hrm.py:
```diff
@@ -0,0 +1,432 @@
+                outputs = model(input_ids=input_ids)
+
+            self.assertIsNotNone(outputs)
+            self.assertIsNotNone(outputs.logits)
```
Same comment

molbap reviewed (commented):
Left some comments, ping me for a full review!

zbloss commented on tests/models/hrm/test_modeling_hrm.py:
```diff
@@ -671,34 +511,73 @@ def test_pretrained_model_reasoning(self):
+            f"Model only got {correct_predictions.item():.1%} of empty cells correct. "
+            f"This suggests the model is not performing meaningful reasoning. "
+            f"Expected at least 50% accuracy on empty cells.",
+        )
```
Yep  this was code left over from some copy-paste work on my part. Removing and updating the comments

zbloss commented on src/transformers/models/hrm/modular_hrm.py:
```diff
@@ -573,42 +522,52 @@ def __init__(self, config: HrmConfig):
+                    "rope_scaling": None,
+                },
+            )()
+            self.rotary_embedding = LlamaRotaryEmbedding(rope_config)
```
Nope, good catch. Removed

zbloss commented on src/transformers/models/hrm/modular_hrm.py:
```diff
@@ -363,94 +335,71 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:
+    # Softmax and apply to values
+    attn_weights = torch.softmax(scores, dim=-1, dtype=torch.float32).to(query.dtype)
+    attn_weights = F.dropout(attn_weights, p=dropout, training=module.training)
+    output = torch.matmul(attn_weights, value)
```
Yep, this is now a straight forward eager attention forward pass, no need to duplicate the codebase.

zbloss committed:
removing duplicate code where possible

zbloss committed:
fix ruff issues

zbloss committed:
add missing imports to fix tests

fix slicing in forward pass

zbloss committed:
fix embeddings test failures

ArthurZucker committed:
Merge branch 'main' into main

ArthurZucker committed:
Merge branch 'main' into main

vasqu commented:
We can close this, HRM has been merged in #46025
